packages/neutrino/Neutrino.js

```diff
@@ -205,39 +150,55 @@ module.exports = class Neutrino {
-      // If middleware is a string, it's a module to require.
-      // Require it, then run the results back through .use()
-      // with the provided options
-      this.use(requireFromRoot(middleware, this.options.root), options);
-    } else if (Array.isArray(middleware)) {
-      // If middleware is an array, it's a pair of some other
-      // middleware type and options
-      this.use(...middleware);
-    } else if (isPlainObject(middleware)) {
-      if ('env' in middleware) {
+      if (middleware.length > 1) {
         throw new ConfigurationError(
-          'Using "env" in middleware has been removed. Apply middleware conditionally instead.'
+          'As of Neutrino 9, middleware only accepts a single argument\n' +
+            'referencing the Neutrino API. Please check that the correct\n' +
+            'value is being passed.'
         );
       }
-      // If middleware is an object, it could contain other middleware in
-      // its "use" property. Run every item in "use" prop back through .use(),
-      // plus set any options.
-      if (middleware.options) {
-        this.options = this.mergeOptions(this.options, middleware.options);
-      }
 
-      if (middleware.use) {
-        if (Array.isArray(middleware.use)) {
-          middleware.use.map(usage => this.use(usage));
-        } else {
-          this.use(middleware.use);
-        }
+      const extraneous = middleware(this);
+
+      if (extraneous && typeof extraneous === 'function') {
+        throw new ConfigurationError(
+          'Neutrino received middleware that upon usage tried returning\n' +
+            'another function. This typically indicates that the supplied\n' +
+            'function should be executed and passed to Neutrino, e.g.\n\n' +
+            '  use: [middleware] -> use: [middleware()]\n' +
+            '  neutrino.use(middleware) -> neutrino.use(middleware())\n\n' +
+            'Please check that the correct value is being passed.'
+        );
       }
+    } else if (typeof middleware === 'string') {
+      throw new ConfigurationError(
+        `"${middleware}" is specified as a string, but as of Neutrino 9,\n` +
+          'middleware can only be passed as functions.\n' +
+          'Use the migration tool and see the migration guide for details:\n' +
+          '  https://neutrinojs.org/migrate\n' +
+          '  https://neutrinojs.org/migration-guide'
+      );
+    } else if (Array.isArray) {
+      throw new ConfigurationError(
+        `"${middleware[0]}" is specified as an array, but as of Neutrino 9,\n` +
+          'middleware can only be passed as functions.\n' +
+          'Use the migration tool and see the migration guide for details:\n' +
+          '  https://neutrinojs.org/migrate\n' +
+          '  https://neutrinojs.org/migration-guide'
+      );
+    } else {
+      throw new ConfigurationError(
+        'As of Neutrino 9, middleware can only be passed as functions.\n' +
```

Comment:
I think the only time this case will be hit is when someone passes something completely invalid (eg an integer, an object or `true`). As such, perhaps the error message does not need to mention Neutrino 9 or the migration guide (since this would have been invalid under Neutrino 8 or earlier too) but instead a more generic "invalid middleware passed, must be a function ..." type message?

Reply:
My concern here was the object case, as I did find a couple of those in the tests, and it was technically valid middleware. I will still remove the migration bits.